lib/shopify_cli/theme/dev_server/proxy.rb

```diff
@@ -112,21 +115,19 @@ def reconstruct_header_name(name)
         end
 
         def build_replace_templates_param(env)
-          params = {}
-
           # Core doesn't support replace_templates
-          return params if @core_endpoints.include?(env["PATH_INFO"])
+          return @templates_to_replace if @core_endpoints.include?(env["PATH_INFO"])
 
           pending_templates = @syncer.pending_updates.select do |file|
```

Comment:
I fail to see the difference between `@syncer.pending_updates` & `@templates_to_replace`.

I agree `@templates_to_replace` needs to be cleaned up at some point, but that what `@syncer.pending_updates` is already doing. It is cleaned up when the file is done uploading.

I think what you're looking for can be achieved by adding a "leeway" (a delay of some sort) before removing  the file from `@syncer.pending_updates` here: https://github.com/Shopify/shopify-cli/blob/87fe16adbc217e4dc7289526c64684300bcf9162/lib/shopify_cli/theme/syncer.rb#L229

Would that fix the issue you think? I would avoid introducing more calls to the backend if possible.

Reply:
To be clear I think `@syncer.pending_updates` is too tightly coupled to `Proxy`, it's a bad design (my fault!).

I'm just worried this fix is a patch on top of my bad design decision from the past.

Reply:
Thanks for this suggestion, @macournoyer! This approach would fix the as well, and we would avoid changes in the `Proxy`.

But, I've got [this idea](https://github.com/Shopify/shopify-cli/pull/1860#issuecomment-995898357) based on your comments, which would avoid the introduction of the delay in the `Syncer` :)